Control.cpp

```diff
@@ -0,0 +1,492 @@
+    desire = desire - initial_difference;
+  }
+  
+  return desire;
+}
+
+double Control::getError(const double current_angle, const double desire){
+  double error = desire - current_angle;
+  
+  if (error < -180) {
+     error += 360;
+   }
+    
+  else if (error > 180) {
+    error -= 360;
+  }
+  
+  return error;
+}
+
+double Control::getErrorUltrasonic(const double current_distance, const double desire) {
+  double error = current_distance - desire;
+  return error;
+}
+
+void Control::getPwm(double &vel) { 
+  if (vel < kLimit_inf_pwm) {
+  vel = kLimit_inf_pwm;
+  }
+  
+  if (vel > kLimit_sup_pwm) {
+    vel = kLimit_sup_pwm;
+  }
+}
+
+double Control::getDistanceFrontLeft() {
+  delay(50);
+  return sonar[0].ping_cm();
+}
+
+double Control::getDistanceFrontRight() {
+  delay(50);
+  return sonar[1].ping_cm();
+}
+
+double Control::getDistanceRightUp() {
+  delay(50);
+  return sonar[2].ping_cm();
+}
+
+double Control::getDistanceRightDown() {
+  delay(50);
+  return sonar[3].ping_cm();
+}
+
+double Control::getDistanceLeftUp() {
+  delay(50);
+  return sonar[4].ping_cm();
+}
+
+double Control::getDistanceLeftDown() {
+  delay(50);
+  return sonar[5].ping_cm(); 
+}
+
+bool Control::checkWallsRight() {
+  double current_distance_right_up = getDistanceRightUp();
+  double current_distance_right_down = getDistanceRightDown();
+  if (current_distance_right_up < 15 and current_distance_right_down < 15) {
+    return true;
+  }
+  else if (current_distance_right_up > 15 || current_distance_right_down > 15) {
+    return false;
+  }
+}
+
+bool Control::checkWallsLeft() {
+  double current_distance_left_up = getDistanceLeftUp();
+  double current_distance_left_down = getDistanceLeftDown();
+  if (current_distance_left_up < 15 and current_distance_left_down < 15) {
+    return true;
+  }
+  else if (current_distance_left_up > 15 || current_distance_left_down > 15) {
+    return false;
+  }
+}
+
+void Control::advance(const double desire, const double desireUltrasonic) {
+  double encoder1 = 0;
+  double encoder2 = 0;
+  double pwm_left = 0;
+  double pwm_right = 0;
+  double final_error = 0;
+  double current_distance_right_up = getDistanceRightUp();
+  double current_distance_right_down = getDistanceRightDown();
+  double current_distance_left_up = getDistanceLeftUp();
+  double current_distance_left_down = getDistanceLeftDown();
+  double error_right_up = getErrorUltrasonic(current_distance_right_up, desireUltrasonic);
+  double error_right_down = getErrorUltrasonic(current_distance_right_down, desireUltrasonic);
+  double error_left_up = getErrorUltrasonic(current_distance_left_up, desireUltrasonic);
+  double error_left_down = getErrorUltrasonic(current_distance_left_down, desireUltrasonic);
+  double current_angle = getAngleBNOX();
+  double errorBNO = getError(current_angle, desire);
+
+  do {
+    double error_ultrasonic_right = error_right_up + error_right_down;
+    double error_ultrasonic_left = error_left_up + error_left_down;
```

Comment:
Are you sure about this? Can you explain? (Reply to this comment if possible)

Reply:
Sumar el error de los dos ultrasonicos de cada lado y juntarlos en una variable, y asi mismo no estar trabajando con cada sensor individualmente

Reply:
Y en las lineas de abajo esta una logica que implemente para los sensores del lado izquierdo, ya que los valores que me daba los sensores del lado derecho es que si el error me da positivo (20(distancia actual) - 15(deseado)), le voy a meter velocidad a los motores del lado izquierdo para que se acerque mas al deseado, y simplemente le cambie el signo al error de los ultrasonicos del lado izquierdo para que fueran iguales y meterle el mismo error a los motores sin que afecte los signos

Reply:
Are you sure adding up the error of the right up sensor and the right down sensor is the best way to do it? I recommend, if you haven't done it yet, to draw a square representing the robot on a hallway (so, with two walls on its side) And draw two more with the robot tilted slightly to the right and to the left, and check what each sensor would give you (more or less than the original one).

And also, what with the cases when there's no wall on one side? Or when the wall is only being sensed by the right_up and not the right_down? Or vice versa